### PR TITLE
codebreaker gem

### PR DESCRIPTION
**Pull request by BogdanPerekhrest34**
done codebreaker gem by https://docs.google.com/document/d/1VW3Mk1WpGkq0FadPih689_k971Zy8inzk6UCPHDLzs/view#

### CODE REVIEW

File: spec/codebreaker_gem/codebreaker_spec.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+require 'rspec'
+require 'codebreaker'
+
+RSpec.describe Codebreaker::Game do
```
Comment (nico-saj) [inline spec/codebreaker_gem/codebreaker_spec.rb:6]: U have to add tests for all examples from 'Codebreaker Examples Matrix'

**Review by nico-saj — CHANGES_REQUESTED** [review]

**Review by nico-saj — CHANGES_REQUESTED** [review]

File: lib/codebreaker/game.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require_relative 'save_result_module'
+
+module Codebreaker
+  DIFFICULTY = {
```
Comment (nico-saj) [inline lib/codebreaker/game.rb:6]: Don't set constants out of class. Move them to Game class

File: spec/codebreaker_gem/codebreaker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    end
+  end
+  context 'codebreaker example input' do
+    [
+      ['6543', '5643', %w[- - + +]], ['6543', '6411', ['+', '-', ' ', ' ']],
+      ['1234', '1234', %w[+ + + +]], ['1234', '1524', ['+', ' ', '-', '+']],
+      ['1234', '3124', %w[- - - +]], ['6543', '2666', [' ', '-', '-', '-']],
+      ['6543', '6544', %w[+ + + -]], ['6666', '1661', [' ', '+', '+', ' ']],
+      ['6543', '3456', %w[- - - -]], ['6543', '2222', [' ', ' ', ' ', ' ']],
+      ['6543', '6666', %w[+ - - -]]
```
Comment (nico-saj) [inline spec/codebreaker_gem/codebreaker_spec.rb:36]: If secret code = 6543 and user code = 6666, response have to be '+'. U can find it in `Codebreaker Examples Matrix`

**Review by nico-saj — CHANGES_REQUESTED** [review]
Recheck logic and fix it. I can't check further PRs while ur game has incorrect logic!

File: spec/codebreaker_gem/codebreaker_spec.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  context 'hint Validation' do
+    it 'incorrect validation' do
+      game.hint
+      game.hint
+      expect { game.hint }.to raise_error(StandardError)
+    end
+  end
+  context 'codebreaker example input' do
+    [
+      ['6543', '5643', %w[- - + +]], ['6543', '6411', ['+', '-', ' ', ' ']],
```
Comment (nico-saj) [inline spec/codebreaker_gem/codebreaker_spec.rb:31]: `['6543', '5643', %w[+ + - -]]` and `['6543', '6411', %w[+ -]]` 
Also use the same style and remove empty strings.

**Review by nico-saj — CHANGES_REQUESTED** [review]

File: lib/codebreaker/save_result_module.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+require 'yaml'
+
+module SaveResultModule
```
Comment (nico-saj) [inline lib/codebreaker/save_result_module.rb:5]: U can't save score in gem and load it in console. All these logic have to be in gem.

File: lib/codebreaker/game.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      @player = player
+      @hints = []
+      @already_check = []
+    end
+
+    def guess_secret_number(input)
+      raise StandardError, 'length = 4; number between 1 and 6' if input.match(/^[1-6]{4}$/).nil?
+
+      input.chars.map(&:to_i).each.with_index do |value, index|
+        next if @output_number[index] == '+'
```
Comment (nico-saj) [inline lib/codebreaker/game.rb:24]: Why do u need this?

File: spec/codebreaker_gem/codebreaker_spec.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    end
+  end
+  context 'codebreaker example input' do
+    [
+      ['6543', '5643', %w[+ + - -]],  ['6543', '2222', []],
+      ['6543', '6411', %w[+ -]],      ['6666', '1661', %w[+ +]],
+      ['6543', '6544', %w[+ + +]],    ['1234', '3124', %w[+ - - -]],
+      ['6543', '3456', %w[- - - -]],  ['1234', '1524', %w[+ + -]],
+      ['6543', '6666', %w[+]],        ['1234', '1234', %w[+ + + +]],
+      ['6543', '2666', %w[-]]
```
Comment (nico-saj) [inline spec/codebreaker_gem/codebreaker_spec.rb:36]: Add one more case: ['1234', '4444', %w[+]]

**Review by nico-saj — CHANGES_REQUESTED** [review]

File: lib/codebreaker/game.rb (hunk 6 above)
Comment (BogdanPerekhrest34) [reply lib/codebreaker/game.rb:24]: what exactly?

File: lib/codebreaker/save_result_module.rb (hunk 5 above)
Comment (nico-saj) [inline lib/codebreaker/save_result_module.rb:5]: U forgot about `module Codebreaker`

File: lib/codebreaker/save_result_module.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  def save_result(full_path = nil)
+    full_path ||= File.expand_path('./data', __dir__)
+    full_path_file = "#{full_path}/score.yml"
```
Comment (nico-saj) [inline lib/codebreaker/save_result_module.rb:8]: Move filename to constant

File: lib/codebreaker/loader_module.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Codebreaker
+  module LoaderModule
+    def load_results(full_path = nil)
+      full_path ||= File.expand_path('./data/score.yml', __dir__)
```
Comment (nico-saj) [inline lib/codebreaker/loader_module.rb:6]: Move filename to constant

**Review by nico-saj — APPROVED** [review]

**Pull request merged**